src/graph.rs

```diff
@@ -288,32 +376,87 @@ impl ModuleGraph {
     Ok(())
   }
 
+  /// Return a vector of references to module objects in the graph. Only modules
+  /// that were fully resolved are present, as "errors" are omitted. If you need
+  /// to know what errors are in the graph, use the `.errors()` method, or if
+  /// you just need to check if everything is "ok" with the graph, use the
+  /// `.valid()` method.
+  #[cfg(feature = "rust")]
+  pub fn modules(&self) -> Vec<&Module> {
+    self
+      .module_slots
+      .iter()
+      .filter_map(|(_, ms)| match ms {
+        ModuleSlot::Module(m) => Some(m),
+        _ => None,
+      })
+      .collect()
+  }
+
   /// Resolve a specifier from the module graph following any possible redirects
   /// returning the "final" module.
   pub fn resolve(&self, specifier: &ModuleSpecifier) -> ModuleSpecifier {
     let mut redirected_specifier = specifier;
+    let mut seen = HashSet::new();
+    seen.insert(redirected_specifier);
     while let Some(specifier) = self.redirects.get(redirected_specifier) {
+      if !seen.insert(specifier) {
+        eprintln!("An infinite loop of redirections detected.\n  Original specifier: {}", specifier);
+        break;
+      }
       redirected_specifier = specifier;
+      if seen.len() > 5 {
+        eprintln!("An excessive number of redirections detected.\n  Original specifier: {}", specifier);
+        break;
+      }
     }
     redirected_specifier.clone()
   }
 
   /// Resolve a dependency of a referring module providing the string specifier
-  /// of the depdency and returning an optional fully qualified module
+  /// of the dependency and returning an optional fully qualified module
   /// specifier.
+  ///
+  /// The `prefer_types` flags indicates if a type dependency is preferred over
+  /// a code dependency. If `true`, a type dependency will be returned in favor
+  /// of a code dependency. If `false` a code dependency will be returned in
+  /// favor of a type dependency. The value should be set to `true` when
+  /// resolving specifiers for type checking, or otherwise `false`.
   pub fn resolve_dependency(
     &self,
     specifier: &str,
     referrer: &ModuleSpecifier,
+    prefer_types: bool,
```

Comment:
Kind of makes me sad when there's a boolean parameter and I can't tell what it is from the call site in a code review. Just makes it easier not to catch a mistake on a code review, but this is a failing of code review tools.

I wonder if it would make sense if we change this to (I'm just throwing some ideas out there, not really suggestions):

```rust
.resolve_dependency_prefer_types(specifier, referrer)
// then a corresponding
.resolve_dependency_prefer_code(specifier, referrer)
```

Or maybe even a struct that's returned that returns both... then people could call:

```rust
.resolve_dependency(specifier, referrer).prefer_types()
```

Reply:
I will think about it, but it is explicit and clear in an editor.

I am not convinced about this suggestion though, having long duplicated methods names or a complicated return structure doesn't make it simple to me.  The compiler would make anyone be explicit about what they want when consuming the API, even if their editor didn't make it clear, which it does.